File: backend/app/db/schema/_skill_sleep.py
"""DDL for Skill-Sleep runs (SkillOpt integration, migration 160).

One row per gated skill-optimization attempt: the blind judge scored the
CURRENT skill body (arm A) vs a CANDIDATE body (arm B) on a held-out question
partition, and the candidate was accepted only if it strictly improved. The
row records both composite scores, the verdict, and provenance for the
operator-adopt surface.
"""

from __future__ import annotations


def create_skill_sleep_tables(conn) -> None:
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS skill_sleep_runs (
            id               INTEGER PRIMARY KEY AUTOINCREMENT,
            project_id       TEXT    NOT NULL,
            skill_name       TEXT    NOT NULL,
            skill_id         INTEGER,
            status           TEXT    NOT NULL DEFAULT 'running'
                CHECK (status IN ('running','accepted','rejected','abstained','failed')),
            current_score    REAL,
            candidate_score  REAL,
            delta            REAL,
            question_count   INTEGER NOT NULL DEFAULT 0,
            partition_seed   INTEGER NOT NULL DEFAULT 0,
            judge_backend    TEXT,
            candidate_body   TEXT,
            reason           TEXT,
            created_at       TEXT    NOT NULL DEFAULT (datetime('now')),
            finished_at      TEXT,
            adopted_at       TEXT
        )
        """
    )
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_skill_sleep_runs_project "
        "ON skill_sleep_runs(project_id, created_at DESC)"
    )
    conn.execute(
        "CREATE INDEX IF NOT EXISTS idx_skill_sleep_runs_skill "
        "ON skill_sleep_runs(skill_name, created_at DESC)"
    )

File: frontend/src/services/api/types/skill-sleep.ts
/**
 * Skill-Sleep types — SkillOpt-style skill optimization runs.
 *
 * A run gates a candidate `SKILL.md` against the current one (blind judge,
 * strict improvement) and measures it on a disjoint split. Accepted candidates
 * are STAGED — `adopted_at` stays null until an operator adopts (the only
 * action that writes the candidate to disk).
 */

export type SkillSleepStatus =
  | 'accepted'
  | 'rejected'
  | 'abstained'
  | 'failed'
  | 'no_candidate';

/** Verdict returned by the `sleep` / `sleep/round` endpoints. */
export interface SkillSleepVerdict {
  run_id: number | null;
  status: SkillSleepStatus;
  accepted: boolean;
  current_score: number | null;
  candidate_score: number | null;
  delta: number | null;
  question_count: number;
  reason: string | null;
  /** Disjoint-split outcome block when measured; shape is backend-owned. */
  outcome?: unknown;
}

/** A persisted `skill_sleep_runs` row (the GET response shape). */
export interface SkillSleepRun {
  id: number;
  project_id: string;
  skill_name: string;
  skill_id: string | null;
  status: SkillSleepStatus;
  current_score: number | null;
  candidate_score: number | null;
  delta: number | null;
  question_count: number;
  partition_seed: number | null;
  judge_backend: string | null;
  candidate_body: string | null;
  /** The current SKILL.md body the candidate was gated against (for the diff);
   *  null on pre-migration-164 runs → drawer falls back to candidate-only. */
  current_body: string | null;
  current_body_hash: string | null;
  reason: string | null;
  created_at: string;
  finished_at: string | null;
  /** The adopt gate hinges on this being null (un-adopted). */
  adopted_at: string | null;
  outcome_before_score: number | null;
  outcome_after_score: number | null;
  outcome_delta: number | null;
  outcome_question_count: number | null;
}

export interface SkillSleepRunsResponse {
  runs: SkillSleepRun[];
}

export interface SkillSleepAdoptResponse {
  adopted: boolean;
  run_id: number;
  reason?: string;
}
